server/sonar-server/src/main/java/org/sonar/server/component/index/ComponentIndexer.java

```diff
@@ -58,44 +62,43 @@ public ComponentIndexer(DbClient dbClient, EsClient esClient) {
    * <b>Warning</b>: This should only be called on an empty index. It does not delete anything.
    */
   public void index() {
-    try (DbSession dbSession = dbClient.openSession(false)) {
-      BulkIndexer bulk = new BulkIndexer(esClient, INDEX_COMPONENTS);
-      bulk.setLarge(true);
-      bulk.start();
-      dbClient.componentDao()
-        .selectAll(dbSession, context -> {
-          ComponentDto dto = (ComponentDto) context.getResultObject();
-          bulk.add(newIndexRequest(toDocument(dto)));
-        });
-      bulk.stop();
-    }
+    doIndexByProjectUuid(null);
   }
 
   /**
    * Update the index for one specific project. The current data from the database is used.
    */
   public void indexByProjectUuid(String projectUuid) {
+    requireNonNull(projectUuid);
+    deleteComponentsByProjectUuid(projectUuid);
+    doIndexByProjectUuid(projectUuid);
+  }
+
+  private void doIndexByProjectUuid(@Nullable String projectUuid) {
+    BulkIndexer bulk = new BulkIndexer(esClient, INDEX_COMPONENTS);
+    bulk.setLarge(true);
```

Comment:
The attribute `setLarge` must be enabled only during server startup because it disables replicas